oozie/src/main/java/org/apache/falcon/workflow/engine/OozieWorkflowEngine.java

```diff
@@ -1434,34 +1434,40 @@ private Frequency getDelay(Feed entity, CoordinatorJob coord) {
     }
 
     private String updateInternal(Entity oldEntity, Entity newEntity, Cluster cluster, BundleJob oldBundle,
-        String user, Boolean skipDryRun) throws FalconException {
+                                  String user, Boolean skipDryRun) throws FalconException {
+        String currentUser = CurrentUser.getUser();
+        switchUser(user);
+
         String clusterName = cluster.getName();
 
         Date effectiveTime = getEffectiveTime(cluster, newEntity);
         LOG.info("Effective time " + effectiveTime);
+        try {
+            //Validate that new entity can be scheduled
+            dryRunForUpdate(cluster, newEntity, effectiveTime, skipDryRun);
+
+            boolean suspended = BUNDLE_SUSPENDED_STATUS.contains(oldBundle.getStatus());
+
+            //Set end times for old coords
+            updateCoords(clusterName, oldBundle, EntityUtil.getParallel(oldEntity), effectiveTime, newEntity);
+            //schedule new entity
+            String newJobId = scheduleForUpdate(newEntity, cluster, effectiveTime, user);
```

Comment:
The function scheduleForUpdate also performs user switch (current user -> user -> current user). However, since you already switch the user before calling scheduleForUpdate, the switch is no longer needed in scheduleForUpdate. Better to remove the "user" argument and user switch in scheduleForUpdate, as this function is only being used in updateInternal.